### DISCUSSION TITLE
How does ref counting work?

### DISCUSSION THREAD

**Discussion by bailuk** [mapsforge/mapsforge#1365] [Q&A] [answered]
I just don't get this. Maybe someone can explain.

1. A tile gets downloaded. This instantiates a Bitmap: refcount 0
2. Same tile gets put() into a cache: refcount 1
3. Tile stays in cache but its reference is returned (get()): refcount 2
4. After tile was painted into Framebuffer: refcount 1
5. Eventually tile gets removed from cache (purge or oldest): refcount 0

But looking at AndroidBitmap.java a bitmap gets destroyed when refcount is -1
What step did I miss?

1: https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/layer/download/TileDownloadThread.java
2, 3 + 5: https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/layer/cache/InMemoryTileCache.java
4 https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/layer/TileLayer.java
?: https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-android/src/main/java/org/mapsforge/map/android/graphics/AndroidBitmap.java

**Comment by devemux86** [comment, marked as answer]
Interesting question, need to study the code.

Though I wonder if the ref count is needed?
Or it just complicates the library maintenance.

What happens if disable the ref count?
How is the performance and memory?

**Comment by bailuk** [comment]
Looking at AndroidBitmap.java shows that the Android part of Mapfsforge does some Bitmap recycling. The destroyBitmap() function gets called on negative refcount.
```java
    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    protected void destroyBitmap() {
        if (this.bitmap != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
                synchronized (REUSABLE_BITMAPS) {
                    REUSABLE_BITMAPS.add(new SoftReference<android.graphics.Bitmap>(this.bitmap));
                }
            } else {
                this.bitmap.recycle();
            }
            this.bitmap = null;
        }
    }
```

But for me even more important is that [mapsforge-gtk](https://github.com/bailuk/mapsforge-gtk) depends on refcounting (gtk and libcario are C Libraries, so there is no java garbage collection).  Thats why I stumbled over this in the first place. Refcounting with Vector maps work but refcounting with tile download maps is leaking (at least in mapsforge-gtk). 

My guess is that download TileDownloadThread must give away owner ship by calling `bitmap.decrementRefCount();` (after line 85). But it could be some other problem...
